### PR TITLE
Refactor SYNC-3664 [v114] app-services update code

### PR DESCRIPTION
**Pull request by bendk**
[SYNC Jira ticket](https://mozilla-hub.atlassian.net/browse/SYNC-3664)

### Description
App-services versions are now going to change and to follow the moz-central release cycle closer.  For rust-components-swift this means the versions will look like:

  - `113.0.YYYYMMDDMMSS` for nightly releases
  - `113.1.X` for "real" releases.

Modified the update script to only try to use nightly releases.

- Does the script change look okay?  Is there a good way to test this other than just seeing how it runs in practice?
- Does the new versioning scheme seem okay?  The app-services releases will only have 2 components, so I figured we could use the extra component to separate out nightlies from non-nightlies.  I'm happy to work with any scheme that the iOS team wants though.
- Does any code or docs need to change to handle non-nightly releases?  This includes switching from a nightly release to the `XXX.1.0` release at the end of the nightly cycle and also would include bumping that version to `XXX.1.1` if app-services did an uplift release (which has historically been very rare).  AFAIK, those are currently handled by hand and they can continue to be with the new versioning system, but please tell me if I should update anything in this repo for that.
### Pull requests checks where applicable
- [ ] Fill in the three TODOs above (tickets number and description of your work)
- [ ] The PR name follows our [naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Accessibility implemented and tested (minimum Dynamic Text and VoiceOver)
- [ ] Unit tests written and passing
- [ ] Documentation / comments for complex code and public methods

### CODE REVIEW

**Comment by bendk** [comment]
@isabelrios I think we talked about the rapid release plan for app-services, what do you think of the code and workflow changes proposed here?  I'm hoping to get all my PRs ready before the end of the nightly cycle so we can make the swap at the start of the 114 cycle.

**Comment by isabelrios** [comment]
> @isabelrios I think we talked about the rapid release plan for app-services, what do you think of the code and workflow changes proposed here? I'm hoping to get all my PRs ready before the end of the nightly cycle so we can make the swap at the start of the 114 cycle.

Hi @bendk! yes, I remember! The changes look good to me however I have some questions in terms of strategy to have this builds running. 
- Is it going to be one nightly build a day? or more? currently our cron job is scheduled to run daily and compare the version we have and if there is a new version. 
- Do we want to pre-build the app with all nighly versions or just the "real" releases, or both? 

Currently when our github action runs, it creates a PR with the new version, but it does not land on main. So, my concern is that we will have:
- Day 1: main + nightly 1
- Day 2: main + nightly 2 
Not sure if we would need to land each PR so we would have:
- Day 1: main + nighly 1
- Day 2: main + nightly 1 + nightly 2

Lastly, I think we should also pre-build with the "real" release at some point.

I'm asking all this to completely understand what would be expected and so adapt our github actions and workflows. 
Thanks! 

(and sorry for my late reply, I was on PTO and then I'm in work week)

**Comment by bendk** [comment]
Thanks for the response, no problem on the delay.

> * Is it going to be one nightly build a day? or more? currently our cron job is scheduled to run daily and compare the version we have and if there is a new version.

Basically once a day.  I think we'll skip building on weekends.  In theory we might want to manually trigger an extra build, but I doubt it.

> * Do we want to pre-build the app with all nighly versions or just the "real" releases, or both?

Does this mean trying to building the app to test for errors?  I think we should definitely do that for nightly versions before merging since the app-services code could have a breaking change.
  - If we added it by accident, then that should block merging.
  - If we added it on purpose, then we should have already gotten a firefox-ios PR approved to resolve the breakage.  In this case, the nightly PR should fail, then we can merge the other branch in to the nightly branch and test that it resolves the issue.


> * Day 1: main + nightly 1
>  * Day 2: main + nightly 2

I think in this case, you could close the day 1 PR and only merge the day 2 one.  If possible, I'd love to have the PRs auto-merged if they build properly.  What's you're opinion on that?

> Lastly, I think we should also pre-build with the "real" release at some point.

Yes, we should think through the real release process a bit.  I wanted to get the nightlies working first and then tackle it.  My guess is that it should go something like this:
 * In the final week of the nightly cycle, a-s goes into a soft-freeze and only merges bugfixes.
 * The day before the nightly cycle ends, a-s makes it's official release.
 * Somehow you incorporate that official release into your workflow, do a pre-build and any other steps needed.  Any way you want to do this is good with me.

**Comment by isabelrios** [comment]
> > * Day 1: main + nightly 1
> > * Day 2: main + nightly 2
> 
> I think in this case, you could close the day 1 PR and only merge the day 2 one. If possible, I'd love to have the PRs auto-merged if they build properly. What's you're opinion on that?

I'm fine with that, we would need developers and EM opinion too.. cc @jevans-mozilla @nbhasin2 

> > Lastly, I think we should also pre-build with the "real" release at some point.
> 
> Yes, we should think through the real release process a bit. I wanted to get the nightlies working first and then tackle it. My guess is that it should go something like this:
> 
>     * In the final week of the nightly cycle, a-s goes into a soft-freeze and only merges bugfixes.
> 
>     * The day before the nightly cycle ends, a-s makes it's official release.
> 
>     * Somehow you incorporate that official release into your workflow, do a pre-build and any other steps needed.  Any way you want to do this is good with me.

Sounds good to me. I think it would be easier if we add another github action to check the official release cadence. So we have nithly and release checks... 
I will open a ticket on our Jira board to start exploring this.

**Comment by isabelrios** [comment]
This is the [epic](https://mozilla-hub.atlassian.net/browse/MTE-865) and tasks associated to track the work needed for us (Mobile Test Engineering team)

**Review by isabelrios — APPROVED** [review]
LGTM!

**Comment by isabelrios** [comment]
The PR looks good to me if we want to start seeing the new workflow working, but I would like to have devs and / or EM feedback on this change. Asking for review too.

**Comment by ddurst** [comment]
Do we need @rpappalax to sign off on this or can we consider this officially approved?

**Comment by isabelrios** [comment]
> Do we need @rpappalax to sign off on this or can we consider this officially approved?

No need, he is my manager, so same team and with one team member approval this would be enough :)

**Comment by bendk** [comment]
We had a snag because of a taskcluster worker issue, but I think this is all ready to go now.  I'm hoping to merge tomorrow and trigger the first rust-components-swift update over the weekend.  I think that will mean a new nightly PR on Monday, does this plan seem good?

Note: I still haven't automated the rust-components-swift updates.  For the start of the project I'll be manually triggering the nightlies.  I think it's probably better to wait on automation, in case there are issues.

**Comment by isabelrios** [comment]
> We had a snag because of a taskcluster worker issue, but I think this is all ready to go now. I'm hoping to merge tomorrow and trigger the first rust-components-swift update over the weekend. I think that will mean a new nightly PR on Monday, does this plan seem good?
> 
> Note: I still haven't automated the rust-components-swift updates. For the start of the project I'll be manually triggering the nightlies. I think it's probably better to wait on automation, in case there are issues.

Sounds good! Once this PR lands, we will work to keep the github action working also for the official releases

**Comment by nbhasin2** [comment]
@isabelrios any objections if we land this?

**Comment by isabelrios** [comment]
> @isabelrios any objections if we land this?

No, no objections. The only thing is that the github action to pre build the app with newly A-S will not work until we update it to find “official” releases not only nightly versions. But we will start working on that really soon

**Comment by nbhasin2** [comment]
@isabelrios sounds good, going to merge this in

**Comment by bendk** [comment]
Thanks for working through this with me.  I'm going to update the rust-components swift code on Monday and trigger the first nightly update then.  If we see issues, maybe I'll trigger another one, but I don't expect to.

I'm going to start working through the non-nightly release process next week and hopefully we can figure out how to update the github action.  I don't think it should be hard at all.

I have some PTO starting next Wednesday, so I'll wait to automate the nightlies until I get back on 5/4.

**Pull request merged**
